[Sema] Playground transform should also log function and closure parameter values

abertelrud opened the pull request:
Add logging of function and closure parameter values when the playground transform is enabled.

### MOTIVATION
The goal of the optional "playground transform" step in Sema is to instrument the code by inserting calls to `__builtin_log()` and similar log functions, in order to record the execution of the compiled code.  Some IDEs (such as Xcode) enable this transform by passing `-playground` and provide implementations of the logger functions that record information that can then be shown in the IDE.

The playground transform currently logs variable assignments and return statements, but it doesn't log the input parameters received by functions and closures.  Knowing these values can be very useful in order to understand the behaviour of functions and closures.

### CHANGES
- add a `ParameterList` parameter to `InstrumenterSupport::transformBraceStmt()`
   - this is an optional parameter list that, if given, specifies the parameters that should be logged at the start of the brace statement
   - this has to be passed into the function because it comes from the owner of the `BraceStmt`
- adjust _PlaygroundTransform.cpp_ to make use of this list
   - the transform will insert calls to `__builtin_log()` for each of the parameters, in order
- adjust _PCMacro.cpp_ to accept the parameter, though this instrumenter doesn't currently make use of the new information
- add two new unit tests (one for functions and one for closures)
- adjust four existing unit tests to account for the new log calls

### REMARKS
- this is currently guarded by the existing "high performance" option (parameter logging is omitted in that case)
- this is the first of a series of log calls I plan to add to the playground transform, and so it might be useful to control them using some kind of organized option set — I will put up a separate PR for discussion with a suggestion for how we might do that

rdar://104974072

abertelrud committed:
Playground transform should log function and closure parameters

Add logging of function and closure parameters when the playground transform is enabled.

MOTIVATION
The goal of the optional "playground transform" step in Sema is to instrument the code by inserting calls to `__builtin_log()` and similar log functions, in order to record the execution of the compiled code.  Some IDEs (such as Xcode) enable this transform by passing `-playground` and provide implementations of the logger functions that record information that can then be shown in the IDE.

The playground transform currently logs variable assignments and return statements, but it doesn't log the input parameters received by functions and closures.  Knowing these values can be very useful in order to understand the behaviour of functions and closures.

CHANGES
- add a `ParameterList` parameter to `InstrumenterSupport::transformBraceStmt()`
   - this is an optional parameter list that, if given, specifies the parameters that should be logged at the start of the brace statement
   - this has to be passed into the function because it comes from the owner of the `BraceStmt`
- adjust _PlaygroundTransform.cpp_ to make use of this list
   - the transform will insert calls to `__builtin_log()` for each of the parameters, in order
- adjust _PCMacro.cpp_ to accept the parameter, though this instrumenter doesn't currently make use of the new information
- add two new unit tests (one for functions and one for closures)
- adjust four existing unit tests to account for the new log calls

QUESTIONS
- should this be off by default and instead be controllable by a new frontend flag?
   - this is the first of a series of log calls I plan to add to the playground transform, and so it might be useful to control them using some kind of organized option set

rdar://104974072

abertelrud commented:
@swift-ci Please smoke test

abertelrud commented:
Looks like I missed updating tests in the `PCMacro` suite... will do that and rerun the CI.

abertelrud committed:
Update the PCMacro `pc_and_log` test that checks the output of `-pc-macro` and `-playground` at the same time.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Adding @cwakamo for visibility and opinion about whether this should default to off and be controllable via another flag.

abertelrud commented:
The failure on `macOS` is in `IRGen/playground.swift` but this doesn't reproduce locally.  Looks like it needs a declaration of `__builtin_log_with_id`, so I'll add that, though it isn't clear to me that there should be any calls to it in the same code.  But this will get us further and we'll see.

Looks like the reason it didn't fail on Linux and Windows is:
```
// REQUIRES: OS=macosx
// REQUIRES: CPU=x86_64
// REQUIRES: objc_interop
```

abertelrud commented:
Oh right, and that's why it doesn't fail locally, I'm on Apple Silicon.  Will try to repro locally on x86_64.

abertelrud committed:
Update the IRGen `playground` test to include declaration of `__builtin_log_with_id` (and fix some of the parameter names while we're at it).  Although no call to it is generated, compilation fails without it.

abertelrud commented:
I verified that the code in `test/IRGen/playground.swift` does not result in a call to `__builtin_log_with_id()`, so it isn't quite clear to me why a declaration of it is needed.

abertelrud commented:
@swift-ci please smoke test

abertelrud committed:
Only log closure parameters that actually have names (i.e. not `_` in a closure, etc).

abertelrud commented:
Found a problem with the unnamed parameters in closures, e.g. `[1,2,3].map { _ in 1 }`.  Skipping those parameters, since it is a strong indication of disinterest in the value by the user.  Updated the newly added lit test for closures to include an example of this kind of unnamed parameter.

abertelrud commented:
@swift-ci please smoke test

xedin commented:
Same could happen with functions as well if there is no internal label i.e. `func test(_: Int)`, isn’t it? I thought that it’s handled properly that’s why I didn’t comment on the use of `->getName()`

abertelrud committed:
Only emit the function and closure parameters when "high performance playground logging" mode is off.  That mode is specifically intended for minimal output of results.

In the future there should probably be an independent option to control logging of function and closure parameters, but for now it's appropriate to gate it by "high performance".

abertelrud commented:
> Same could happen with functions as well if there is no internal label i.e. `func test(_: Int)`, isn’t it? I thought that it’s handled properly that’s why I didn’t comment on the use of `->getName()`

I see, thank you for pointing that out!  The fix I put in to check for `->hasName()` should work for both, and I will update the unit test to make sure it does.

I also gated this on the flag indicating high performance mode, so that we don't log the parameters in that case.  I would like to propose a follow-on PR to provide more fine-grained control over whether to log parameters, but that can be done separately.

abertelrud committed:
Add a test for unnamed function parameters, and update some comments in the code and tests.

abertelrud commented:
Updated the unit test to also cover `func foo(_: Int)` case.  Thanks @xedin!

abertelrud commented:
@swift-ci please smoke test

abertelrud committed:
Fix comment about unnamed parameters.

abertelrud commented:
@swift-ci please smoke test

xedin commented:
No problem!

abertelrud commented:
Looks like all the tests failed because of some SwiftDriver issue now.  Rerunning.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
I see — there is indeed a `lit` failure here from making this condition on high performance mode being off.  Will fix it.

abertelrud committed:
Revert the change to PCMacro `pc_and_log` test that checks the output of `-pc-macro` and `-playground` at the same time.

The reason is that this test passes `-playground-high-performance` and therefore doesn't see the new changes, now that they are controlled by that option.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
At this point I believe that this PR is completely ready.  @xedin are there others who should review as well, or is this ready to be merged in your view?  Thanks!

xedin commented:
The original reviewer probably, I think it’s ready to go though.

abertelrud commented:
> The original reviewer probably, I think it’s ready to go though.

Great, thanks.  This looks like quite old code and I wasn't able to identify a current "owner" of the code.  I have spoke with @cwakamo about the nature of the changes though, and others affected by it.  Thanks for your reviews!
